### PR TITLE
VxAdmin: Loading state for "Save Election Package"

### LINKED ISSUE #4043: VxAdmin: Add a loading screen to the save ballot package modal
Add a loading screen to the save ballot package modal

Issue automatically created from slack feedback reported by caroline 
https://votingworks.slack.com/archives/CJU9MSC6S/p1696535762719999

### PR DESCRIPTION
**Pull request by jonahkagan**
## Overview

_Review by commit_

Fixes #4043 

- Adds a new `loading` prop to `Button` that disables the button and shows a loading spinner.
- Uses this prop to add a loading state to the Save Election Package modal.

## Demo Video or Screenshot

https://github.com/votingworks/vxsuite/assets/530106/78f61728-92ee-480f-b926-260cbdbbda39

## Testing Plan
Manual and automated tests

## Checklist

- [ ] I have added [logging](https://github.com/votingworks/vxsuite/tree/main/libs/logging) where appropriate to any new user actions, system updates such as file reads or storage writes, or errors introduced.

- [ ] I have added a screenshot and/or video to this PR to demo the change
- [ ] I have added the "user_facing_change" label to this PR to automate an announcement in #machine-product-updates

### CODE REVIEW

File: apps/admin/frontend/src/components/export_election_package_modal_button.tsx
Comment (jonahkagan) [inline apps/admin/frontend/src/components/export_election_package_modal_button.tsx:58]: @votingworks/vxsuite-eng Just a reminder that if you use `.mutateAsync` without a `try/catch`, any exceptions will cause an unhandled promise rejection, which makes tests crash in weird ways.

Recommended ways to call a mutation are:
```ts
// If you don't care about waiting for completion
mutation.mutate(args)

// If you need to wait for completion or results
mutation.mutate(args, { 
  onSuccess: (results) => {
     ... do the next thing ... 
  }
})

// If you need to pass an async event handler to another 
// component, or using a promise will make things
// way more  readable (e.g. multiple mutations in a row)
async function onSomeEvent() {
  try {
    await mutation.mutateAsync(args)
  } catch (e) {
    // Handled by query client default error handling
   }
}
```

File: apps/admin/frontend/src/components/export_election_package_modal_button.tsx
Comment (kofi-q) [inline apps/admin/frontend/src/components/export_election_package_modal_button.tsx:89]: optional: after watching the video again, a nice-to-have here would be to change the text to `Saving...` or something along those lines while the mutation's running

File: apps/admin/frontend/src/components/export_election_package_modal_button.tsx
Comment (kofi-q) [inline apps/admin/frontend/src/components/export_election_package_modal_button.tsx:113]: We're losing the disabled-while-saving state of cancel button here - should we be adding a conditional disable on line 92 when the mutation is running? (And additionally, a no-op in `closeModal()`?)

File: libs/ui/src/button.tsx
Comment (jonahkagan) [reply libs/ui/src/button.tsx:53]: Yeah, I went back and forth. I had initially implemented this with the existing props.

I tried out adding a `loading` prop for two reasons:
- It felt a bit confusing to read, since there were two places that conditioned on the mutation loading state (the `disabled` prop and the `icon` prop)
- Currently we have sporadic disabling of buttons during submit. I think it would be great to standardize on this pattern because we have a lot of submit buttons, so I wanted to make it super simple for other engineers to choose this approach. The UI library we have in Arlo uses a prop like this, and I've found it really useful.

Do either of those reasons feel compelling? I'm open to switching it back if you feel like there's a disadvantage or risk to trade off with those reasons.

File: apps/admin/frontend/src/components/export_election_package_modal_button.tsx
Comment (kofi-q) [reply apps/admin/frontend/src/components/export_election_package_modal_button.tsx:113]: +1 - was concerned enabling a 'cancel' button that doesn't actually cancel the action (and the fact that this state of the modal is step 1 of a 2-part flow, since there's a follow-up UI to eject the drive)

File: libs/ui/src/button.tsx
Comment (kofi-q) [reply libs/ui/src/button.tsx:53]: The API does seem a bit opaque -- unclear at first glance without digging into the `Button` code that this is shorthand for disabling the button and changing the icon, which was a sign that it doesn't quite fit and part of the motivation for keeping the `Button` instances explicit.
(Actually think something like `disabled={isMutating}` is clearer to the reader and not fully convinced that the lack of a feature like the `loading` prop is why people may forget to disable buttons during submit -- they might forget the `loading` prop just as easily.)

That being said, I don't feel strongly enough to block on that, so definitely down with your call on this.

**Review by kofi-q — APPROVED** [review]
LGTM, assuming the cancel button disabling logic's getting added

File: apps/admin/frontend/src/components/export_election_package_modal_button.tsx
Comment (jonahkagan) [reply apps/admin/frontend/src/components/export_election_package_modal_button.tsx:89]: https://github.com/votingworks/vxsuite/assets/530106/77be74e8-2dc3-41a3-985b-3f30f5ca5d39

I like it!

File: libs/ui/src/button.tsx
Comment (jonahkagan) [reply libs/ui/src/button.tsx:53]: That makes sense and I think those are valid concerns.

Once I added the changing of the button label as well, it started to feel less "clean" to use the `loading` prop anyway.

Instead I kind of wanted to do something like this:

```ts
              {saveElectionPackageToUsbMutation.isLoading ? (
                <Button
                  disabled
                  icon="Loading"
                  variant="primary"
                  onPress={() => {}}
                >
                  Saving...
                </Button>
              ) : (
                <Button
                  icon="Export"
                  onPress={saveElectionPackage}
                  variant="primary"
                >
                  Save
                </Button>
              )}
```

This is kind of annoying because you have to add the `onPress` prop to the loading button even though its disabled. So that gave me an idea for a different approach:

```ts
              {saveElectionPackageToUsbMutation.isLoading ? (
                <LoadingButton variant="primary">Saving...</LoadingButton>
              ) : (
                <Button
                  icon="Export"
                  onPress={saveElectionPackage}
                  variant="primary"
                >
                  Save
                </Button>
              )}
```

Here, `LoadingButton` would be a separate component built off of `Button` that implements the pattern of using the loading icon and disabling. I like that you only have to have one conditional in this approach. The downside is that you have to repeat the `variant` prop.

What do you think of this approach?

For completeness of options, here's what it looks like with no new API:

```ts
              <Button
                icon={
                  saveElectionPackageToUsbMutation.isLoading
                    ? 'Loading'
                    : 'Export'
                }
                onPress={saveElectionPackage}
                variant="primary"
                disabled={saveElectionPackageToUsbMutation.isLoading}
              >
                {saveElectionPackageToUsbMutation.isLoading
                  ? 'Saving...'
                  : 'Save'}
              </Button>
```

Comment (kofi-q) [reply libs/ui/src/button.tsx:53]: Excellent -- `<LoadingButton>` approach feels much cleaner for when we want to change more than just the `disabled` state.
Having to repeat the `variant` is a minor inconvenience, compared to what we're getting -- but, shameless plug for keeping the `disabled` state styling consistent across variants, 'casue `variant` wouldn't matter in that case 😜

Comment (jonahkagan) [reply libs/ui/src/button.tsx:53]: Sweet, I'll go for the `LoadingButton` and we can see how we like it.

I think it would be a bit weird for a primary action button to visually become "unprimary" once the action is in progress, so that makes me hesitant to change the disabled state styling to disregard variant.

**Pull request merged**
